Amittal94/ca_to_pva

amittal94 opened the pull request:
A more structured format for converting from pyepics result to pva data types. I also made a validator to set defaults instead of None but need to identify where to use it. I'm fairly sure in p4p unless the whole section i.e alarm is none, then the internal keys are always present. So currently, it should only be needed in ca stuff, that is to say, CAScalarAny, which might be the only use case.

amittal94 committed:
general PyEpics converter

amittal94 committed:
precision set a default value for channel acess

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -1,30 +1,133 @@
+from enum import Enum
+
 import epac.flatbuffers.data_types as dt
+from epac.flatbuffers.fbschemas.pva0.AlarmStatus import AlarmStatus
```
Note: we'll need to make this available at a more public path.

amittal94 replied:
Can you elaborate on this?

stephen-jd-dann replied:
You've had to break an abstraction boundary here IMO. `fbschemas` is an implementation detail of this library. If client code wanted to use `AlarmStatus` it would currently need to use this internal path to import it.

Of course there's the question of whether typical client code would use `AlarmStatus`, but the fact you've done it here suggests to me that the answer to that is "yes".

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -1,30 +1,133 @@
+
+
+class PyEpicsConverter:
+    class CAAlarmStatus(Enum):
```
Style: `CaAlarmStatus` is the preferred option I think. This applies to any acronym.

Suggestion: It would be useful to have a comment saying where you got this from (in other words, citation needed)

Suggestion: if you make this an `IntEnum` you'll be able to compare values to ints, which I think would let you remove all those `.value` below.

amittal94 replied:
I did not think of that, might need to make a a commit later on the whole flatbuffers to update names like this

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -1,30 +1,133 @@
+        READ_ACCESS = 20
+        WRITE_ACCESS = 21
+
+    ALARM_TYPE = {
```
This is a fascinating and surprisingly elegant way of building a repetitive dict.

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -1,30 +1,133 @@
+    UNDEFINED_ALARM = AlarmStatus.UNDEFINED  # UNDEFINED
+
+    @classmethod
+    def create_alarm(cls, severity: int, status: int) -> dt.AlarmT:
```
**Suggestion:**

```suggestion
    def create_alarm(cls, severity: int, ca_status: int) -> dt.AlarmT:
```

In fact it might be worth completely avoiding variables named `status` in this file, and requiring them all to be `pva_status` or `ca_status`.

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -1,30 +1,133 @@
+        """Creates and returns an AlarmT object."""
+        alarm_status = cls.ALARM_TYPE.get(status, cls.UNDEFINED_ALARM)
+        try:
+            alarm_msg = cls.CAAlarmStatus(status).name
```
**Bug:** The message for `NO_ALARM` should be blank (`""`)

amittal94 replied:
I saw a case of the message being "NO_ALARM" via caget and pvget though (might be an implementation choice perhaps)? If the empty string is certain, then I guess I would have to kind of explicitly set for the 0 case.

stephen-jd-dann replied:
Good point, I was only looking at PVXS.

Since it's clearly an implementation choice, we can leave this as-is. If needed we can change it in the future for efficiency reasons.

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -1,30 +1,133 @@
+        try:
+            alarm_msg = cls.CAAlarmStatus(status).name
+        except ValueError:
+            alarm_msg = ""
```
Suggestion:

```suggestion
            alarm_msg = "undefined alarm status"
```

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -1,30 +1,133 @@
-        control=control,
-    )
+    @classmethod
+    def cascalarany_to_ntscalarany(cls, data: dt.CAScalarAny) -> dt.NTScalarAny:
```
Suggestion: a nicer API might be to accept a dictionary too and convert it to `CAScalarAny`

amittal94 replied:
Would that not be hidden behavior? I thought we wanted to be rather explicit with the initial conversion.

stephen-jd-dann replied:
IMO it's not hidden because the whole point of this method is to do conversion from CA to PVA. I just think of it as a nice QoL thing. Otherwise client code will have a lot of boilerplate, where you convert a dict to a `CAScalarAny` and then immediately call this method.

But it's just a thought. It's hard to predict how code will be used.

stephen-jd-dann reviewed (changes requested):
Nice and neat. a couple of required changes I think and a few suggestions.

I also think we could have better naming. `PyEpicsConverter` might be better as `CaToNtConverter` or something and `cascalarany_to_ntscalarany` could probably be `convert`.

Finally, a thought for the future. If we want conversion preferences (e.g. do not include display parameters), one way is to use instance methods, and store the preferences on the object. We should consider making `cascalarany_to_ntscalarany` an instance method to make sure no API changes will be needed.

amittal94 commented:
> I also think we could have better naming. `PyEpicsConverter` might be better as `CaToNtConverter` or something and `cascalarany_to_ntscalarany` could probably be `convert`.

Yeah fair enough, I keep thinking with a mindset of something other than pyepics, but really we will only use one channel access. I will simplify the naming, but a seperate convert will be for waveforms, so maybe convert_scalar, and convert_waveform will do.

stephen-jd-dann commented:
I forgot to say in my review: it would be nice to see at least one unit test for this. It's very unit-testable code.

amittal94 committed:
Refactorisation and Reviewed Changes

- rename PyEpicsConverter to CatoNtConverter
- change CAAlarmStatus to CaAlarmStatus
- change the enum to an intenum
- handle the alarm status using the enum (required for proper typing)
- explicit message for an undefined alarm status
- data -> data.value (bug)
- rename the conversion function to convert_scalar
- modify the conversion function to also accept dictionaries
- rename status to ca_status where appropriate
- access AlarmStatus via datatypes to keep abstraction

amittal94 commented:
What you said about the instance methods could actually be applied to the serialise and deserialise data functions as well potentially later, I can make some sort of class and use set it up to only serialise, deserialise what we require. Internal fields like units could also be potentially handled with some creativity I think.

amittal94 committed:
modify the convert scalar method to optionally ignore fields

amittal94 committed:
set defaults for cascalar

amittal94 committed:
initial unit tests for ca to pva

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -114,20 +114,41 @@ def create_control(limit_low: float, limit_high: float) -> dt.ControlT:
+    def convert_scalar(
+        self,
+        data: dt.CAScalarAny | dict,
+        include_alarm: bool = True,
+        include_timestamp: bool = True,
+        include_display: bool = True,
+        include_control: bool = True,
```
Can you pop a `*,` before these to make them keyword-only? I don't always remember to do that, but IMO it's best practice for these arguments that are confusing if used positionally.

stephen-jd-dann commented on tests/test_converter.py:
```diff
@@ -0,0 +1,64 @@
+        assert result.display.precision == test_data_dict["precision"]
+        assert isinstance(result.control, dt.ControlT)
+        assert result.control.limitLow == test_data_dict["lower_ctrl_limit"]
+        assert result.control.limitHigh == test_data_dict["upper_ctrl_limit"]
```
This whole block would probably be clearer if it was `assert result == expected_result`. You'd have to construct the expected result of course. Although it's tricky if there's some part of it you don't care about.

stephen-jd-dann commented on tests/test_converter.py:
```diff
@@ -0,0 +1,64 @@
+        result_ca_scalar = converter.convert_scalar(test_data_ca_scalar)
+        assert result == result_ca_scalar
+
+    def test_missing_data_defaults(self):
```
At the moment this test is a bit of a "test that the code really is what I already wrote". I think it would be more interesting if it also tested the conversion to PVData.

stephen-jd-dann commented on src/epac/flatbuffers/ca_to_pva.py:
```diff
@@ -114,20 +114,41 @@ def create_control(limit_low: float, limit_high: float) -> dt.ControlT:
-    @classmethod
-    def cascalarany_to_ntscalarany(cls, data: dt.CAScalarAny) -> dt.NTScalarAny:
-        """Converts CAScalarAny to NTScalarAny."""
+    def convert_scalar(
```
Thought: it would be nice if we could automatically exclude e.g. the display if it's absent from the CA data. Defaults are all well and good, but we really ought to say "don't know". Unfortunately, the information about whether a particular field is present is currently lost on conversion to `CAScalarAny`.

I think we leave it for now, but a later PR might want to make most of those fields optional.

amittal94 replied:
I would actually like it optional but its not obvious how to proceed given that we will have limited fields from CA rather than all the ones in the corresponding normative types. Would need some sort of decision along the lines "if there is no status and severity omit alarm. " but then what if we only have status, though this may a long shot edge case.

stephen-jd-dann replied:
That's one of those things that's possible from a data model point of view, but not in practice. Channel access always sends the alarm status and severity together. I think the other fields are in the same situation.

amittal94 replied:
I will keep that in mind then, leaving it for a future pr like you suggested, among other reworks

stephen-jd-dann reviewed (commented):
I'm happy with most of this. Some room for improvement still but nothing that can't be done in future PRs.

amittal94 committed:
make ca_to_pva kws keyword only

amittal94 committed:
more tests for ca to pva
